### ISSUE TITLE
How to design a system with low density?

### ISSUE THREAD

**Issue by daohoangson** [jfkirk/tensorrec#96]
I have read through (most of) the issues and have the recommendation up and running but the result is not as good as I hoped. I have a few questions looking for your answer. Below is info about my current dataset:

```
user_features.shape = (79770, 26035) # avg features/user = 2
item_features.shape = (21093, 20157) # avg features/item = 4
interactions.shape = (79770, 21093)
```

I train with this code, which is similar to your example:

```
model = tensorrec.TensorRec()
model.fit(interactions, user_features, item_features, epochs=1000, verbose=True)
```

However, my hardware is not powerful enough so I had to picked samples from the full dataset instead of running it in full:

```
sample_user_features.shape = (4, 26035)
sample_interactions.shape = (4, 21093)

predicted_ranks = model.predict_rank(user_features=sample_user_features, item_features=item_features)
r_at_k = tensorrec.eval.recall_at_k(predicted_ranks, sample_interactions, k=10)
print(np.mean(r_at_k))
```

The recall@10 value is 0.00006. I manually inspected the predicted item for these users and it's not good. So my questions are:

1. I think my avg features/user of 2 and avg features/item of 4 are too small. Do you think so? Should the recommendation works with such low density?
2. I choose 4 user ids from the dataset and build sample matrix to run the recall. I'm not confident about this step. Is this mathematically and statistically correct?
3. I understand that you have implemented `fit_partial`. In my situation, the item features can increase over time. Will `fit_partial` works with resized matrix?

Thank you.

**Comment by jfkirk** [comment]
Hey @daohoangson ! Thanks for asking about this.

To your questions:
1. I think your average number of features is not the issue. What could be the issue is how they are handled. Are the feature values binary 1/0 or continuous?
2. For a more representative recall sample, you may want to increase your user sample size. In your position, I would increase the sample size as far as my machine were capable of handling. What is the density of your interactions? How many avg. interactions per user?
3. At the moment, TensorRec is unable to handle changes in the number of user/item features. This change would require re-constructing the TensorFlow graph and re-training from scratch.

You may get much better results using a batched loss graph (rather than RMSE which is default). Try using WMRBLossGraph and you may get significantly better results.
```model = TensorRec(loss_graph=WMRBLossGraph())```

**Comment by daohoangson** [comment]
Thank you for getting back to me.

1. All of my features (user+item) are binary (0 and 1)
2. I will try increasing the sample size to see if it improves (will report back the result). The density of the interactions matrix is 0.0013 (2234372 non zeros in the (79770, 21093) matrix, my values are non negative btw)
3. That is understood.

I will also switch to WMRBLossGraph as you suggested and report back. Thank you.

**Comment by daohoangson** [comment]
Hmm, I'm unable to complete the model fitting with WMRBLossGraph. The python process is killed everytime. I have tried n_sampled_items=2/20/200/2000, it always get killed about 30s after "Beginning fitting". Can you give any pointer? @jfkirk?

I'm testing on a VM with 4vCPU/16GB RAM.

**Comment by daohoangson** [comment]
Running again with 16vCPU/32GB RAM: it got further but still getting killed after 2 epochs:

```
INFO:root:Beginning fitting
INFO:root:EPOCH 0 BATCH 0 loss = 9.925631523132324, weight_reg_l2_loss = 0.2309596875, mean_pred = -0.0005950707127340138
INFO:root:EPOCH 1 BATCH 0 loss = 8.303492546081543, weight_reg_l2_loss = 0.18411509765625, mean_pred = 0.6903960108757019
Killed
```

**Comment by daohoangson** [comment]
Thanks for the tip, I will try that.

**Comment by daohoangson** [comment]
I have been able to trim the user features and item features for the model fit to run in complete:

```
user_features.shape = (25419, 908) # avg features/user = 2
item_features.shape = (19484, 9522) # avg features/item = 3
interactions.shape = (25419, 19484) # density 0.0008
epochs= 20
n_sampled_items= 1948
user_batch_size= 100
```

For the evaluation, I picked 1k and 10k users. Ran eval twice and ended up with the same recall@10 of `0.0488`.

I checked the recommendation manually and it appears the items returned are the popular ones (item with lots of interactions) and they are not actually related to user features. This is probably a problem because my avg features/user and avg features/item is too small?

Another issue I noticed while running fitting with `verbose=True`: the reported loss value started at 9 and ended at 9 too. I would assume the algo will minimize loss so it should goes down after 20 epochs right?

**Comment by daohoangson** [comment]
The previous run was done on my laptop. Just got my hand back on the server, I'm running the fit model with user_batch_size=800, the verbose log now looks better. Currently at epoch 2 and loss decreased from 9 -> 7 already.

**Comment by daohoangson** [comment]
So the less value is about 7 at the end of model fitting. And recall@10 is `0.0503`. Apparently using larger `user_batch_size` yields better model?

**Comment by jfkirk** [comment]
@daohoangson The difference between `0.0503` and `0.488` might not be significant enough to say that larger `user_batch_size` yields a better model.

20 epochs may be a bit too short to learn much nuance for the recommendation. I'd recommend increasing it, if that's feasible.

Regarding popularity, I've run in to this same problem. You can try turning `biased=False` when building the model and try using `BalancedWMRBLossGraph`. Both of these have helped with "popularity bias" I've encountered in the past.

Let me know if you try this out!

**Comment by daohoangson** [comment]
I have some interesting data after fitting the model dozen of times:

| epochs | recall@10 for training dataset | recall@10 for test dataset |
| --- | --- | --- |
| 1 | 0.1830187142740457 | 0.3662953988132286 |
| 2 | 0.27352707746481286 | 0.4972308268123855 |
| 3 | 0.3199103676358411 | 0.5631426097188966 |
| 4 | 0.3426408854173336 | 0.5869633480813036 |
| 5 | 0.33708967513935 | 0.5905252102026333 |
| 6 | 0.3532201746788516 | 0.6063555255200556 |
| 7 | 0.3621703240323549 | 0.6171398821760207 |
| 8 | 0.36146319125016696 | 0.6169908706467532 |
| 9 | 0.3657490125767337 | 0.6241734098001693 |
| 10 | 0.366262853110833 | 0.6215061048647799 |
| 20 | 0.3716591071408457 | 0.6223120855316361 |
| 50 | 0.37144562195032504 | 0.6221324166860508 |
| 100 | 0.36904940316788687 | 0.6150610005859628 |
| 500 | 0.36806407978532174 | 0.6211129739832923 |

I splited our full dataset into 2 parts to train/test separately. The train dataset is not as "rich" as the test dataset, that may explain the increased recall for the same model. 

I have run fit/test with `BalancedWMRBLossGraph` exclusively this time and it looked like the result does not get better after the 10th epoch. Maybe this is as good as it may get for our data. Do you think I should try 1k epoch or higher? It's taking 40s per epoch for me so I'm a bit reluctant to it...

**Comment by EricC91** [comment]
Hi @daohoangson,

I am confronted with almost the same results as you...
Did you finally find parameters that work well?

Thank you

**Comment by daohoangson** [comment]
@EricC91 I have had redone my input matrices so it's denser and switched to use `BalancedWMRBLossGraph`. The result is good enough for our use case.
